File: server/sources/socrata.js
// socrata.js — one adapter FAMILY covering every Socrata (SODA) open-data feed. Adding
// a city = adding a row to DATASETS, not writing code. These are official municipal
// open-data portals (public-official, keyless; an optional app token just lifts the
// anonymous rate limit). Each row says where the point is, what the call-type field is,
// and how to classify it.
//
// We order by the system field :updated_at DESC (present on every dataset, so a wrong
// per-dataset column name can't 400 us), pull the most-recent N, then bbox-filter in JS.

import { getJson } from './_http.js';
import { makeEvent, inBbox, bboxIntersects, kindFromText, sevFromText, numOrNull } from './types.js';

// Pull a [lat, lon] out of a Socrata row, tolerating the three shapes these feeds use:
// numeric lat/lon columns, a GeoJSON point column, or a {latitude,longitude} location.
function extractLatLon(r, ds) {
  let la = numOrNull(r[ds.latField]), lo = numOrNull(r[ds.lonField]);
  if (la != null && lo != null) return [la, lo];
  for (const f of [ds.pointField, 'point', 'location', 'report_location', 'intersection_point', 'geocoded_column', 'the_geom']) {
    const p = f && r[f];
    if (!p) continue;
    if (Array.isArray(p.coordinates) && p.coordinates.length >= 2) return [numOrNull(p.coordinates[1]), numOrNull(p.coordinates[0])];
    if (p.latitude != null && p.longitude != null) return [numOrNull(p.latitude), numOrNull(p.longitude)];
  }
  return [null, null];
}

async function fetchSocrata(ds, bbox, cfg) {
  if (!bboxIntersects(bbox, ds.region)) return [];
  const params = new URLSearchParams();
  params.set('$limit', String(ds.limit || 400));
  params.set('$order', ':updated_at DESC');
  const headers = cfg?.socrataToken ? { 'X-App-Token': cfg.socrataToken } : {};
  const url = `https://${ds.host}/resource/${ds.dataset}.json?${params.toString()}`;
  const rows = await getJson(url, { headers });
  const out = [];
  for (const r of rows) {
    const [la, lo] = extractLatLon(r, ds);
    if (la == null || lo == null || !inBbox(bbox, la, lo)) continue;
    const typeText = ds.typeField ? (r[ds.typeField] || '') : '';
    const kind = ds.kind || kindFromText(typeText);
    const severity = ds.severity != null ? ds.severity : sevFromText(typeText);
    const ts = ds.tsField ? Date.parse(r[ds.tsField]) : Date.now();
    const nativeId = (ds.idField && r[ds.idField]) || `${la.toFixed(5)},${lo.toFixed(5)}`;
    const ev = makeEvent({
      source: ds.id, nativeId, kind, severity, lat: la, lon: lo,
      title: typeText || ds.label,
      description: r[ds.addrField] || r.address || r.incident_address || r.street_address || '',
      sourceUrl: `https://${ds.host}/resource/${ds.dataset}`,
      ts: Number.isFinite(ts) ? ts : Date.now(),
      raw: r,
    });
    if (ev) out.push(ev);
  }
  return out;
}

// region = a rough city bbox so we never call a city's portal when the observer is
// nowhere near it. kind/severity: omit to classify from the call-type text.
const DATASETS = [
  { id: 'sea-fire-cad', label: 'Seattle Fire/EMS dispatch', host: 'data.seattle.gov', dataset: 'kzjm-xkqj',
    latField: 'latitude', lonField: 'longitude', typeField: 'type', tsField: 'datetime', idField: 'incident_number', addrField: 'address',
    kinds: ['fire', 'medical'], attribution: 'Seattle Fire · data.seattle.gov',
    region: { minLat: 47.4, maxLat: 47.78, minLon: -122.46, maxLon: -122.22 } },

  { id: 'sf-pd-cad', label: 'SF Police dispatch (real-time)', host: 'data.sfgov.org', dataset: 'gnap-fj3t',
    pointField: 'intersection_point', typeField: 'call_type_final_desc', tsField: 'received_datetime', idField: 'cad_number', addrField: 'intersection_name',
    kinds: ['police'], kind: 'police', attribution: 'SFPD · data.sfgov.org',
    region: { minLat: 37.70, maxLat: 37.84, minLon: -122.54, maxLon: -122.34 } },

  { id: 'sf-fire-cad', label: 'SF Fire calls', host: 'data.sfgov.org', dataset: 'nuek-vuh3',
    pointField: 'case_location', typeField: 'call_type', tsField: 'received_dttm', idField: 'call_number', addrField: 'address',
    kinds: ['fire', 'medical'], attribution: 'SF Fire · data.sfgov.org',
    region: { minLat: 37.70, maxLat: 37.84, minLon: -122.54, maxLon: -122.34 } },

  { id: 'chi-311', label: 'Chicago 311 service requests', host: 'data.cityofchicago.org', dataset: 'v6vf-nfxy',
    latField: 'latitude', lonField: 'longitude', typeField: 'sr_type', tsField: 'created_date', idField: 'sr_number', addrField: 'street_address',
    kind: 'civic', severity: 0, kinds: ['civic'], attribution: 'City of Chicago · data.cityofchicago.org',
    region: { minLat: 41.62, maxLat: 42.05, minLon: -87.95, maxLon: -87.52 } },

  { id: 'chi-crime', label: 'Chicago crimes (7-day lag)', host: 'data.cityofchicago.org', dataset: 'ijzp-q8t2',
    latField: 'latitude', lonField: 'longitude', typeField: 'primary_type', tsField: 'date', idField: 'case_number', addrField: 'block',
    kind: 'police', kinds: ['police'], attribution: 'Chicago PD · data.cityofchicago.org',
    region: { minLat: 41.62, maxLat: 42.05, minLon: -87.95, maxLon: -87.52 } },

  { id: 'nyc-311', label: 'NYC 311 service requests', host: 'data.cityofnewyork.us', dataset: 'erm2-nwe9',
    latField: 'latitude', lonField: 'longitude', typeField: 'complaint_type', tsField: 'created_date', idField: 'unique_key', addrField: 'incident_address',
    kind: 'civic', severity: 0, kinds: ['civic'], attribution: 'NYC OpenData · data.cityofnewyork.us',
    region: { minLat: 40.48, maxLat: 40.93, minLon: -74.27, maxLon: -73.68 } },

  { id: 'cin-cad', label: 'Cincinnati PD dispatch', host: 'data.cincinnati-oh.gov', dataset: 'gexm-h6bt',
    latField: 'latitude_x', lonField: 'longitude_x', typeField: 'incident_type_desc', tsField: 'create_time_incident', idField: 'event_number', addrField: 'address_x',
    kinds: ['police'], attribution: 'Cincinnati PD · data.cincinnati-oh.gov',
    region: { minLat: 39.05, maxLat: 39.32, minLon: -84.72, maxLon: -84.25 } },
];

export default DATASETS.map((ds) => ({
  id: ds.id,
  category: 'incidents',
  kinds: ds.kinds || ['civic'],
  keyless: true,
  attribution: ds.attribution,
  label: ds.label,
  enabled: () => true,
  fetch: (bbox, cfg) => fetchSocrata(ds, bbox, cfg).catch(() => []),
}));

File: src-tauri/src/proxy/sources/socrata.rs
// Socrata (SODA) municipal open-data family — native mirror of server/sources/socrata.js.
// Adding a city = adding a row to DATASETS. Ordered by the :updated_at system field (present
// on every dataset, so a wrong per-dataset column can't 400 us), then bbox-filtered locally.

use serde_json::Value;

use super::{get_json, kind_from_text, make_event, num, s_of, sev_from_text, ts_or_now, Bbox};
use crate::server::AppState;

pub struct Ds {
    pub id: &'static str,
    pub label: &'static str,
    pub host: &'static str,
    pub dataset: &'static str,
    pub lat_field: &'static str,
    pub lon_field: &'static str,
    pub point_field: &'static str,
    pub type_field: &'static str,
    pub ts_field: &'static str,
    pub id_field: &'static str,
    pub addr_field: &'static str,
    pub kind: &'static str, // "" -> classify from the call-type text
    pub severity: i64,      // <0 -> classify from the call-type text
    pub region: (f64, f64, f64, f64),
}

pub const DATASETS: &[Ds] = &[
    Ds { id: "sea-fire-cad", label: "Seattle Fire/EMS dispatch", host: "data.seattle.gov", dataset: "kzjm-xkqj",
         lat_field: "latitude", lon_field: "longitude", point_field: "", type_field: "type", ts_field: "datetime",
         id_field: "incident_number", addr_field: "address", kind: "", severity: -1,
         region: (47.4, 47.78, -122.46, -122.22) },
    Ds { id: "sf-pd-cad", label: "SF Police dispatch (real-time)", host: "data.sfgov.org", dataset: "gnap-fj3t",
         lat_field: "", lon_field: "", point_field: "intersection_point", type_field: "call_type_final_desc",
         ts_field: "received_datetime", id_field: "cad_number", addr_field: "intersection_name", kind: "police", severity: -1,
         region: (37.70, 37.84, -122.54, -122.34) },
    Ds { id: "sf-fire-cad", label: "SF Fire calls", host: "data.sfgov.org", dataset: "nuek-vuh3",
         lat_field: "", lon_field: "", point_field: "case_location", type_field: "call_type",
         ts_field: "received_dttm", id_field: "call_number", addr_field: "address", kind: "", severity: -1,
         region: (37.70, 37.84, -122.54, -122.34) },
    Ds { id: "chi-311", label: "Chicago 311 service requests", host: "data.cityofchicago.org", dataset: "v6vf-nfxy",
         lat_field: "latitude", lon_field: "longitude", point_field: "", type_field: "sr_type", ts_field: "created_date",
         id_field: "sr_number", addr_field: "street_address", kind: "civic", severity: 0,
         region: (41.62, 42.05, -87.95, -87.52) },
    Ds { id: "chi-crime", label: "Chicago crimes (7-day lag)", host: "data.cityofchicago.org", dataset: "ijzp-q8t2",
         lat_field: "latitude", lon_field: "longitude", point_field: "", type_field: "primary_type", ts_field: "date",
         id_field: "case_number", addr_field: "block", kind: "police", severity: -1,
         region: (41.62, 42.05, -87.95, -87.52) },
    Ds { id: "nyc-311", label: "NYC 311 service requests", host: "data.cityofnewyork.us", dataset: "erm2-nwe9",
         lat_field: "latitude", lon_field: "longitude", point_field: "", type_field: "complaint_type", ts_field: "created_date",
         id_field: "unique_key", addr_field: "incident_address", kind: "civic", severity: 0,
         region: (40.48, 40.93, -74.27, -73.68) },
    Ds { id: "cin-cad", label: "Cincinnati PD dispatch", host: "data.cincinnati-oh.gov", dataset: "gexm-h6bt",
         lat_field: "latitude_x", lon_field: "longitude_x", point_field: "", type_field: "incident_type_desc",
         ts_field: "create_time_incident", id_field: "event_number", addr_field: "address_x", kind: "", severity: -1,
         region: (39.05, 39.32, -84.72, -84.25) },
];

/// Tolerate the three shapes these feeds use: numeric lat/lon columns, a GeoJSON point
/// column, or a Socrata {latitude, longitude} location object.
fn extract_lat_lon(r: &Value, ds: &Ds) -> Option<(f64, f64)> {
    if !ds.lat_field.is_empty() {
        if let (Some(la), Some(lo)) = (r.get(ds.lat_field).and_then(num), r.get(ds.lon_field).and_then(num)) {
            return Some((la, lo));
        }
    }
    for f in [ds.point_field, "point", "location", "report_location", "intersection_point", "geocoded_column", "the_geom"] {
        if f.is_empty() {
            continue;
        }
        if let Some(p) = r.get(f) {
            if let Some(c) = p.get("coordinates").and_then(|c| c.as_array()) {
                if c.len() >= 2 {
                    if let (Some(lo), Some(la)) = (num(&c[0]), num(&c[1])) {
                        return Some((la, lo));
                    }
                }
            }
            if let (Some(la), Some(lo)) = (p.get("latitude").and_then(num), p.get("longitude").and_then(num)) {
                return Some((la, lo));
            }
        }
    }
    None
}

pub async fn fetch(st: &AppState, ds: &Ds, b: &Bbox) -> Result<Vec<Value>, String> {
    if !b.intersects(ds.region) {
        return Ok(vec![]);
    }
    let url = format!(
        "https://{}/resource/{}.json?$limit=400&$order=:updated_at%20DESC",
        ds.host, ds.dataset
    );
    let rows = get_json(st, &url, "application/json").await?;
    let arr = match rows.as_array() {
        Some(a) => a,
        None => return Ok(vec![]),
    };
    let src_url = format!("https://{}/resource/{}", ds.host, ds.dataset);
    let mut out = Vec::new();
    for r in arr {
        let (la, lo) = match extract_lat_lon(r, ds) {
            Some(x) => x,
            None => continue,
        };
        if !b.contains(la, lo) {
            continue;
        }
        let type_text = if ds.type_field.is_empty() { "" } else { s_of(r, ds.type_field) };
        let kind = if ds.kind.is_empty() { kind_from_text(type_text) } else { ds.kind };
        let sev = if ds.severity >= 0 { ds.severity } else { sev_from_text(type_text) };
        let ts = ts_or_now(r.get(ds.ts_field));
        let id_val = s_of(r, ds.id_field);
        let native = if id_val.is_empty() { format!("{:.5},{:.5}", la, lo) } else { id_val.to_string() };
        let title = if type_text.is_empty() { ds.label } else { type_text };
        let desc = s_of(r, ds.addr_field);
        if let Some(ev) = make_event(ds.id, &native, kind, sev, la, lo, title, desc, Value::from(src_url.clone()), ts, Value::Null) {
            out.push(ev);
        }
    }
    Ok(out)
}
